### ISSUE TITLE
Implement SteamAudio as an available audio library

### ISSUE THREAD

**Issue by gabeklavans** [panda3d/panda3d#1619]
## Description
The [Steam Audio SDK](https://valvesoftware.github.io/steam-audio/index.html) was just [fully open sourced](https://steamcommunity.com/games/596420/announcements/detail/7745698166044243233). This is a modern, actively-developed audio library designed by Valve specifically for games and VR, and could prove to be a powerful alternative to the default OpenAL-Soft library for Panda3D.

## Use Case
Steam Audio most likely supports all the things that Panda3D currently implements with OpenAL, and may have the benefit of a simpler, more modern API with features more tailored to 3D games specifically. I haven't done an extensive comparison myself ([here is a list of the features in the SteamAudio C API](https://valvesoftware.github.io/steam-audio/doc/capi/reference.html)), but there are some discussions floating on it by people more knowledgeable than me, such as [here](https://www.reddit.com/r/pcgaming/comments/1av1a6a/steam_audio_open_source_release/).
It may not be worth the effort, as OpenAL-Soft seems to meet Panda3D's needs, but I figured I'd start the dialogue here.

**Comment by serkkz** [comment]
If I understand the terms of the Apache 2.0 license correctly, the developer must distribute his application under this license, that is, as an open source. Which is an imposition of licensing, I don't think Panda3D users will be thrilled.

**Comment by gabeklavans** [comment]
> If I understand the terms of the Apache 2.0 license correctly, the developer must distribute his application under this license, that is, as an open source. Which is an imposition of licensing, I don't think Panda3D users will be thrilled.

I'm not a licensing expert but the breakdown on this page seems to suggest that you don't need to apply the Apache 2.0 license to projects that use software under that license https://fossa.com/blog/open-source-licenses-101-apache-license-2-0/

**Comment by serkkz** [comment]
https://github.com/ValveSoftware/steam-audio/blob/master/LICENSE.md

As indicated here in Part 4, paragraph "c", it is required to keep the copyright notice form. The problem is that the author of the work can simply include a mention of copyright at the beginning of the source code. Accordingly, you will be forced to distribute it further in this form.

Maybe I'm exaggerating, but that's how I understand it.

**Comment by gabeklavans** [comment]
> https://github.com/ValveSoftware/steam-audio/blob/master/LICENSE.md
> 
> As indicated here in part 4, paragraph "c", it is required to keep the copyright notice form. The problem is that the author of the work can simply include a mention of copyright at the beginning of the source code. Accordingly, you will be forced to distribute it further in this form.
> 
> Maybe I'm exaggerating, but that's how I understand it.

My understanding is that applies to a derivative of SteamAudio itself, not any project that uses a binary from the unmodified source. But that's definitely at or beyond the limit of my understanding, so I'll let someone more qualified/experienced figure that out

**Comment by serkkz** [comment]
Yes, that's right, in fact, you are allowed to change the license at your discretion with compatibility. However, I also don't understand what to do with the SteamAudio source code.

**Comment by Wizzerinus** [comment]
apache 2.0 is not a copyleft license, so that part shouldn't be an issue on that front

**Comment by gabeklavans** [comment]
> However, I also don't understand what to do with the SteamAudio source code.

From my understating, SteamAudio could replace OpenAL-Soft as an HRTF provider. SteamAudio also provides lots of effects processing for modeled and simulated 3D environments, presumably with high efficiency and with a good API. Of course those last two points are to be actually investigated, first, but it seems promising.

Also I'm not sure what sort of VR support Panda3D has but SteamAudio seems to tout its VR-capabilities.

**Comment by gabeklavans** [comment]
I suppose SteamAudio going open source doesn't actually affect what I'm proposing, but it makes it more of a pretty option for inclusion, just for being open source, but also since it could be modified for Panda3D's needs and avoid any sort of potential proprietary licensing needs.

**Comment by rdb** [comment]
Certainly seems worth implementing.  The Apache license is not a problem.

**Comment by lifelandman** [comment]
I don't know if anyone's working on this already and I doubt I have the skills (yet) to tackle this, but were steam audio implemented, would it share the same interface as panda3d's currently-implemented sound libraries? Considering how many features it has that rely on ray-testing geometry provided to the api, I somewhat struggle to imagine what that would look like.
